From be5661825b1172d55c190a087ceb8907187d523a Mon Sep 17 00:00:00 2001
From: Ben Beasley <code@musicinmybrain.net>
Date: Thu, 15 May 2025 08:14:53 -0700
Subject: [PATCH] PR #1888: Adjust Table.GrowExtremelyLargeTable to avoid OOM
 on i386

Imported from GitHub PR https://github.com/abseil/abseil-cpp/pull/1888

While this only covers `i386`/`i686`, which is the motivation for this PR, this test can be expected to OOM on any 32-bit platform. For now, this is the minimal change that avoids the problem [in the Fedora package](https://src.fedoraproject.org/rpms/abseil-cpp/).

This fixes one of the two test failures reported in https://github.com/abseil/abseil-cpp/issues/1887.
Merge 395acb74da05fa35d924059a170ebd8267356b65 into f28774a28227c4e30041616bff4aa0120ed724c0

Merging this change closes #1888

COPYBARA_INTEGRATE_REVIEW=https://github.com/abseil/abseil-cpp/pull/1888 from musicinmybrain:extremely-large-table-32-bit 395acb74da05fa35d924059a170ebd8267356b65
PiperOrigin-RevId: 759154889
Change-Id: I0a105fc42c51898c277b4a056ccd6599b43e1a50
---
 absl/container/internal/raw_hash_set_test.cc | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/absl/container/internal/raw_hash_set_test.cc b/absl/container/internal/raw_hash_set_test.cc
index a5cbd44d3b2..2c55d0fc079 100644
--- a/absl/container/internal/raw_hash_set_test.cc
+++ b/absl/container/internal/raw_hash_set_test.cc
@@ -4267,8 +4267,8 @@ struct ConstUint8Hash {
 // 5. Finally we will catch up and go to overflow codepath.
 TEST(Table, GrowExtremelyLargeTable) {
   constexpr size_t kTargetCapacity =
-#if defined(__wasm__) || defined(__asmjs__)
-      NextCapacity(ProbedItem4Bytes::kMaxNewCapacity);  // OOMs on WASM.
+#if defined(__wasm__) || defined(__asmjs__) || defined(__i386__)
+      NextCapacity(ProbedItem4Bytes::kMaxNewCapacity);  // OOMs on WASM, 32-bit.
 #else
       NextCapacity(ProbedItem8Bytes::kMaxNewCapacity);
 #endif
